pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml

```diff
@@ -0,0 +1,228 @@
+
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Subclass of List</description>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+        public class MyClass {
+          void loop(ArrayList<String> l) {
+            for (int i = 0; i < l.size(); i++) {
+              System.out.println(l.get(i));
+
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Get called on another list</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+        public class MyClass {
+          void loop(List<String> l) {
+            List<String> l2 = new ArrayList<>(l);
+            for (int i = 0; i < l.size(); i++) {
+              System.out.println(l2.get(i));
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Backwards iteration</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+        public class MyClass {
+          void loop(List<String> l) {
+            for (int i = l.size() - 1; i > 0; i-= 1) {
+              System.out.println(i + ": " + l.get(i));
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+
+    <test-code>
+        <description>Index var initialized outside for init</description>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+        public class MyClass {
+          void loop(List<String> l) {
+            int i = 0;
+            for (; i < l.size(); i++) {
+              System.out.println(l.get(i));
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+
+    <test-code>
+        <description>Array positives</description>
+        <expected-problems>2</expected-problems>
+        <code><![CDATA[
+        class Foo {
+          protected static final char[] filter(char[] chars, char removeChar) {
+            int count = 0;
+            for (int i = 0; i < chars.length; i++) {
+              if (chars[i] == removeChar) {
+                count++;
+              }
+            }
+
+            char[] results = new char[chars.length - count];
+
+            int index = 0;
+            for (int i = 0; i < chars.length; i++) {
+              if (chars[i] != removeChar) {
+                results[index++] = chars[i];
+              }
+            }
+            return results;
+          }
+        }
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Consider iterators</description>
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+        class Foo {
+          void loop() {
+            for (Iterator<DataFlowNode> i = path.iterator(); i.hasNext();) {
+              DataFlowNode inode = i.next();
+              if (inode.getVariableAccess() == null) {
+                continue;
```

Comment:
if we had this same example, but where doing a `i.remove()` here, then the code wouldn't be possible to rewrite as a foreach loop, but the rule would flag it non the less

Reply:
Nope, the iterator only has those 3 methods: `remove`, `hasNext` and `next`,  and being able to use `remove` without a `ConcurrentModificationException` is pretty much the only valid reason to use an iterator instead of a foreach